src/unicode/fold.c

```diff
@@ -0,0 +1,135 @@
+ *
+ *    As a special exception, the copyright holders give permission to link the
+ *    code of portions of this program with the OpenSSL library under certain
+ *    conditions as described in each individual source file and distribute
+ *    linked combinations including the program with the OpenSSL library. You
+ *    must comply with the Server Side Public License in all respects for
+ *    all of the code used other than as permitted herein. If you modify file(s)
+ *    with this exception, you may extend this exception to your version of the
+ *    file(s), but you are not obligated to do so. If you do not wish to do so,
+ *    delete this exception statement from your version. If you delete this
+ *    exception statement from all source files in the program, then also delete
+ *    it in the license file.
+ *
+ *    THIS IS A GENERATED FILE, DO NOT MODIFY.
+ */
+
+#include "fold.h"
+#include "../mc-range-edge-generation-private.h" // For mc_count_leading_zeros_u64
+#include "../mongocrypt-private.h"
+
+static bool append_utf8_codepoint(bson_unichar_t codepoint, char **output_it, mongocrypt_status_t *status) {
+    if (codepoint <= 0x7f /* max 1-byte codepoint */) {
+        *(*output_it)++ = (char)codepoint;
+    } else if (codepoint <= 0x7ff /* max 2-byte codepoint*/) {
+        *(*output_it)++ = (char)((codepoint >> (6 * 1)) | 0xc0); // 2 leading 1s.
+        *(*output_it)++ = (char)(((codepoint >> (6 * 0)) & 0x3f) | 0x80);
+    } else if (codepoint <= 0xffff /* max 3-byte codepoint*/) {
+        *(*output_it)++ = (char)((codepoint >> (6 * 2)) | 0xe0); // 3 leading 1s.
+        *(*output_it)++ = (char)(((codepoint >> (6 * 1)) & 0x3f) | 0x80);
+        *(*output_it)++ = (char)(((codepoint >> (6 * 0)) & 0x3f) | 0x80);
+    } else {
+        if (codepoint > 0x10FFFF) {
+            CLIENT_ERR("append_utf8_codepoint: codepoint was out of range for UTF-8");
+            return false;
+        }
+        *(*output_it)++ = (char)((codepoint >> (6 * 3)) | 0xf0); // 4 leading 1s.
+        *(*output_it)++ = (char)(((codepoint >> (6 * 2)) & 0x3f) | 0x80);
+        *(*output_it)++ = (char)(((codepoint >> (6 * 1)) & 0x3f) | 0x80);
+        *(*output_it)++ = (char)(((codepoint >> (6 * 0)) & 0x3f) | 0x80);
+    }
+    return true;
+}
+
+// C translation of mongo::unicode::String::caseFoldAndStripDiacritics.
+bool unicode_fold(const char *str,
+                  size_t len,
+                  unicode_fold_options_t options,
+                  char **out_str,
+                  size_t *out_len,
+                  mongocrypt_status_t *status) {
+    BSON_ASSERT_PARAM(str);
+    BSON_ASSERT_PARAM(out_str);
+    BSON_ASSERT_PARAM(out_len);
+    BSON_ASSERT_PARAM(status);
+
+    if (!(options & (kUnicodeFoldRemoveDiacritics | kUnicodeFoldToLower))) {
+        CLIENT_ERR("unicode_fold: Either case or diacritic folding must be enabled");
+        return false;
+    }
+    *out_str = bson_malloc(len);
```

Comment:
Following [comment](https://github.com/10gen/mongo/blob/f35c66a4417ab546443730bba3549cf53a314ba1/src/mongo/db/fts/unicode/string.cpp#L171-L174), allocate space to account for cases where folding increases UTF-8 encoded length:

```suggestion
    // Allocate space for possible growth. Folding characters may result in longer UTF-8 sequences.
    // 2x is an upper bound. With current fold maps, the largest growth is a 2-byte sequence mapping to a 3-byte sequence.
    *out_str = bson_malloc(2 * len);
```

The true maximum growth appears to be 3/2: https://gist.github.com/kevinAlbs/bd4d916f62fa3a49f07bc1c6a679b67a

Add a test to confirm:

```c
TEST_UNICODE_FOLD("\xe2\xb1\xa6", 4, "\xc8\xbe", 3, kUnicodeFoldToLower);
```

Reply:
Thanks so much for this! To give some background -- I originally constructed the two folding maps such that any cases where byte length would increase were excluded, so I changed this so that we only allocate the exact size of the buffer. But later, I realized that the algorithm in the paper as it was written can actually handle the byte-length-increasing case as long as the codepoint length doesn't increase, so I reconstructed the maps without fixing this. Reading your comment, I realized I actually may have to change this back, since in the new iteration of the paper I believe increasing byte length is also bad. Confirming with the crypto team on this.

Reply:
Changed and added a test.